collector/src/main.rs

```diff
@@ -1141,7 +1166,100 @@ fn main_result() -> anyhow::Result<i32> {
+                DownloadSubcommand::Git { url } => download_from_git(&target_dir, &url)?,
+                DownloadSubcommand::Crate { krate, version } => {
+                    download_from_crates_io(&target_dir, &krate, &version)?
+                }
+            };
+            println!("Benchmark stored at {}", target_dir.display());
+            Ok(0)
+        }
+    }
+}
+
+fn check_target_dir(target_dir: &Path, overwrite: bool) -> anyhow::Result<()> {
+    if target_dir.exists() {
+        if overwrite {
+            std::fs::remove_dir_all(&target_dir).expect(&format!(
+                "Cannot remove previous directory at {}",
+                target_dir.display()
+            ));
+        } else {
+            return Err(anyhow::anyhow!(
+                "Directory {} already exists",
+                target_dir.display()
+            ));
+        }
     }
+    Ok(())
+}
+
+fn get_downloaded_crate_target(benchmark_dir: &Path, cmd: &DownloadCommand) -> PathBuf {
+    let name = cmd.name.clone().unwrap_or_else(|| match &cmd.command {
+        DownloadSubcommand::Git { url } => url
+            .trim_end_matches("/")
+            .trim_end_matches(".git")
+            .split("/")
+            .last()
+            .expect("Crate name could not be determined from git URL")
+            .to_string(),
+        DownloadSubcommand::Crate { krate, version } => format!("{krate}-{version}"),
+    });
+    PathBuf::from(benchmark_dir).join(name)
+}
+
+fn download_from_git(target: &Path, url: &str) -> anyhow::Result<()> {
+    let tmpdir = tempfile::TempDir::new().unwrap();
+    Command::new("git")
+        .arg("clone")
+        .arg(url)
+        .arg(tmpdir.path())
+        .status()
+        .expect("Git clone failed");
+    generate_lockfile(tmpdir.path());
+    execute::rename(&tmpdir, &target)?;
+    Ok(())
+}
+
+fn download_from_crates_io(target_dir: &Path, krate: &str, version: &str) -> anyhow::Result<()> {
+    let url = format!("https://crates.io/api/v1/crates/{krate}/{version}/download");
+    let response = reqwest::blocking::get(url)
+        .expect("Cannot download crate")
+        .error_for_status()?;
+
+    let data = flate2::read::GzDecoder::new(response);
+    let mut archive = tar::Archive::new(data);
+
+    let tmpdir = tempfile::TempDir::new().unwrap();
+    archive.unpack(&tmpdir)?;
+    let unpacked_dir = std::fs::read_dir(&tmpdir.path())
+        .unwrap()
+        .next()
+        .unwrap()
+        .unwrap()
+        .path();
+    generate_lockfile(&unpacked_dir);
+    execute::rename(&unpacked_dir, &target_dir)?;
+
+    Ok(())
+}
+
+fn generate_lockfile(directory: &Path) {
+    let manifest_path = directory.join("Cargo.toml");
+
+    // Cargo metadata should do nothing if there is already a lockfile present.
+    // Otherwise it will generate a lockfile.
+    Command::new("cargo")
+        .arg("metadata")
+        .current_dir(manifest_path.parent().unwrap())
+        .stdout(std::process::Stdio::null())
```

Comment:
We should also pipe stderr to null here -- at least as far as I can tell, Cargo will still emit a warning about passing --format-version here to stderr.

Or, maybe what makes more sense, is to pass --format-version 1 -- that'll prevent the warning but keep any download/fetching messages Cargo may print which seems nice.